// nativescript-audio/android/player.js
import { Application, Observable, Utils } from '@nativescript/core';
import { resolveAudioFilePath } from '../common';
import { AudioPlayerEvents } from '../options';
export var AudioFocusDurationHint;
(function (AudioFocusDurationHint) {
    AudioFocusDurationHint[AudioFocusDurationHint["AUDIOFOCUS_GAIN"] = android.media.AudioManager.AUDIOFOCUS_GAIN] = "AUDIOFOCUS_GAIN";
    AudioFocusDurationHint[AudioFocusDurationHint["AUDIOFOCUS_GAIN_TRANSIENT"] = android.media.AudioManager
        .AUDIOFOCUS_GAIN_TRANSIENT] = "AUDIOFOCUS_GAIN_TRANSIENT";
    AudioFocusDurationHint[AudioFocusDurationHint["AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK"] = android.media.AudioManager
        .AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK] = "AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK";
    AudioFocusDurationHint[AudioFocusDurationHint["AUDIOFOCUS_GAIN_TRANSIENT_EXCLUSIVE"] = android.media.AudioManager
        .AUDIOFOCUS_GAIN_TRANSIENT_EXCLUSIVE] = "AUDIOFOCUS_GAIN_TRANSIENT_EXCLUSIVE";
})(AudioFocusDurationHint || (AudioFocusDurationHint = {}));
const defaultAudioFocusManagerOptions = {
    durationHint: AudioFocusDurationHint.AUDIOFOCUS_GAIN,
    usage: android.media.AudioAttributes.USAGE_MEDIA,
    contentType: android.media.AudioAttributes.CONTENT_TYPE_MUSIC
};
export class AudioFocusManager extends Observable {
    constructor(options) {
        super();
        this._mAudioFocusGranted = false;
        this._audioPlayerSet = new Set();
        this._mOnAudioFocusChangeListener = new android.media.AudioManager.OnAudioFocusChangeListener({
            onAudioFocusChange: (focusChange) => {
                this.notify({
                    eventName: 'audioFocusChange',
                    object: this,
                    focusChange
                });
            }
        });
        options = Object.assign(Object.assign({}, defaultAudioFocusManagerOptions), (options || {}));
        this._durationHint = options.durationHint;
        if (android.os.Build.VERSION.SDK_INT < 26) {
            return;
        }
        const playbackAttributes = new android.media.AudioAttributes.Builder()
            .setUsage(options.usage)
            .setContentType(options.contentType)
            .build();
        this._audioFocusRequest = new android.media.AudioFocusRequest.Builder(options.durationHint)
            .setAudioAttributes(playbackAttributes)
            .setAcceptsDelayedFocusGain(true)
            .setOnAudioFocusChangeListener(this._mOnAudioFocusChangeListener)
            .build();
    }
    needsFocus() {
        return this._audioPlayerSet.size > 0;
    }
    requestAudioFocus(owner) {
        let result = true;
        let focusResult = null;
        if (!this._mAudioFocusGranted) {
            const ctx = this._getAndroidContext();
            const am = ctx.getSystemService(android.content.Context.AUDIO_SERVICE);
            if (android.os.Build.VERSION.SDK_INT >= 26) {
                focusResult = am.requestAudioFocus(this._audioFocusRequest);
            }
            else {
                focusResult = am.requestAudioFocus(this._mOnAudioFocusChangeListener, android.media.AudioManager.STREAM_MUSIC, this._durationHint);
            }
            if (focusResult === android.media.AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                result = true;
            }
            else {
                result = false;
            }
        }
        this._audioPlayerSet.add(owner);
        this._mAudioFocusGranted = result;
        return result;
    }
    abandonAudioFocus(owner) {
        if (owner) {
            if (!this._audioPlayerSet.has(owner)) {
                return this._mAudioFocusGranted;
            }
            this._audioPlayerSet.delete(owner);
        }
        if (this.needsFocus() || !this._mAudioFocusGranted) {
            return this._mAudioFocusGranted;
        }
        const ctx = this._getAndroidContext();
        const am = ctx.getSystemService(android.content.Context.AUDIO_SERVICE);
        let result = null;
        if (android.os.Build.VERSION.SDK_INT >= 26) {
            console.log('abandonAudioFocusRequest...', this._audioFocusRequest);
            result = am.abandonAudioFocusRequest(this._audioFocusRequest);
            console.log('abandonAudioFocusRequest...result...', result);
        }
        else {
            console.log('abandonAudioFocus...', this._mOnAudioFocusChangeListener);
            result = am.abandonAudioFocus(this._mOnAudioFocusChangeListener);
        }
        if (result === android.media.AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            this._mAudioFocusGranted = false;
        }
        else {
            console.log('Failed to abandon audio focus.');
        }
        return this._mAudioFocusGranted;
    }
    _getAndroidContext() {
        let ctx = Application.android.context;
        if (!ctx) {
            ctx = Application.getNativeApplication().getApplicationContext();
        }
        if (ctx === null) {
            setTimeout(() => {
                this._getAndroidContext();
            }, 200);
            return;
        }
        return ctx;
    }
}
let globalMixingManager;
function getGlobalMixingManager() {
    if (!globalMixingManager) {
        globalMixingManager = new AudioFocusManager();
    }
    return globalMixingManager;
}
export class TNSPlayer {
    constructor(durationHint = AudioFocusDurationHint.AUDIOFOCUS_GAIN) {
        this._wasPlaying = false;
        if (!(durationHint instanceof AudioFocusManager)) {
            this.setAudioFocusManager(new AudioFocusManager({
                durationHint: durationHint
            }));
        }
        else {
            this.setAudioFocusManager(durationHint);
        }
    }
    get events() {
        if (!this._events) {
            this._events = new Observable();
        }
        return this._events;
    }
    get android() {
        return this._player;
    }
    get volume() {
        const ctx = this._getAndroidContext();
        const mgr = ctx.getSystemService(android.content.Context.AUDIO_SERVICE);
        return mgr.getStreamVolume(android.media.AudioManager.STREAM_MUSIC);
    }
    set volume(value) {
        if (this._player && value >= 0) {
            this._player.setVolume(value, value);
        }
    }
    get duration() {
        if (this._player) {
            return this._player.getDuration();
        }
        else {
            return 0;
        }
    }
    get currentTime() {
        return this._player ? this._player.getCurrentPosition() : 0;
    }
    setAudioFocusManager(manager) {
        var _a, _b, _c;
        if (manager === this._audioFocusManager) {
            return;
        }
        (_a = this._audioFocusManager) === null || _a === void 0 ? void 0 : _a.off('audioFocusChange', this._onAudioFocusChange, this);
        (_b = this._audioFocusManager) === null || _b === void 0 ? void 0 : _b.abandonAudioFocus(this);
        this._audioFocusManager = manager;
        (_c = this._audioFocusManager) === null || _c === void 0 ? void 0 : _c.on('audioFocusChange', this._onAudioFocusChange, this);
    }
    initFromFile(options) {
        return new Promise((resolve, reject) => {
            options.autoPlay = false;
            this.playFromFile(options).then(resolve, reject);
        });
    }
    playFromFile(options) {
        return new Promise((resolve, reject) => {
            try {
                this._options = options;
                if (options.autoPlay !== false) {
                    options.autoPlay = true;
                }
                const audioPath = resolveAudioFilePath(options.audioFile);
                this._player.setAudioStreamType(android.media.AudioManager.STREAM_MUSIC);
                this._player.reset();
                this._player.setDataSource(audioPath);
                if (Utils.isFileOrResourcePath(audioPath)) {
                    this._player.prepare();
                }
                else {
                    this._player.prepareAsync();
                }
                if (options.infoCallback) {
                    this._player.setOnInfoListener(new android.media.MediaPlayer.OnInfoListener({
                        onInfo: (player, info, extra) => {
                            options.infoCallback({ player, info, extra });
                            return true;
                        }
                    }));
                }
                this._player.setOnPreparedListener(new android.media.MediaPlayer.OnPreparedListener({
                    onPrepared: mp => {
                        if (options.autoPlay) {
                            this.play();
                        }
                        resolve(null);
                    }
                }));
            }
            catch (ex) {
                this._abandonAudioFocus();
                reject(ex);
            }
        });
    }
    initFromUrl(options) {
        return new Promise((resolve, reject) => {
            options.autoPlay = false;
            this.playFromUrl(options).then(resolve, reject);
        });
    }
    playFromUrl(options) {
        return new Promise((resolve, reject) => {
            resolve(this.playFromFile(options));
        });
    }
    pause() {
        return new Promise((resolve, reject) => {
            try {
                if (this._player && this._player.isPlaying()) {
                    this._player.pause();
                    this._abandonAudioFocus(true);
                    this._sendEvent(AudioPlayerEvents.paused);
                }
                resolve(true);
            }
            catch (ex) {
                reject(ex);
            }
        });
    }
    play() {
        return new Promise((resolve, reject) => {
            var _a;
            try {
                console.log('player play()');
                if (this._player && !this._player.isPlaying()) {
                    if (this._options.audioMixing) {
                        this.setAudioFocusManager(getGlobalMixingManager());
                    }
                    const audioFocusGranted = this._requestAudioFocus();
                    if (!audioFocusGranted) {
                        throw new Error('Could not request audio focus');
                    }
                    this._sendEvent(AudioPlayerEvents.started);
                    Application.android.foregroundActivity.setVolumeControlStream(android.media.AudioManager.STREAM_MUSIC);
                    Application.android.registerBroadcastReceiver(android.media.AudioManager.ACTION_AUDIO_BECOMING_NOISY, (context, intent) => {
                        this.pause();
                    });
                    if ((_a = this._options) === null || _a === void 0 ? void 0 : _a.pitch) {
                        const playBackParams = new android.media.PlaybackParams();
                        playBackParams.setPitch(this._options.pitch);
                        this._player.setPlaybackParams(playBackParams);
                    }
                    this._player.start();
                }
                resolve(true);
            }
            catch (ex) {
                reject(ex);
            }
        });
    }
    resume() {
        if (this._player) {
            this.play();
            this._sendEvent(AudioPlayerEvents.started);
        }
    }
    seekTo(time) {
        return new Promise((resolve, reject) => {
            try {
                if (this._player) {
                    time = time * 1000;
                    this._player.seekTo(time);
                    this._sendEvent(AudioPlayerEvents.seek);
                }
                resolve(true);
            }
            catch (ex) {
                reject(ex);
            }
        });
    }
    changePlayerSpeed(speed) {
        var _a, _b;
        if (android.os.Build.VERSION.SDK_INT >= 23 && this.play) {
            if ((_a = this._player) === null || _a === void 0 ? void 0 : _a.isPlaying()) {
                this._player.setPlaybackParams(this._player.getPlaybackParams().setSpeed(speed));
            }
            else {
                this._player.setPlaybackParams(this._player.getPlaybackParams().setSpeed(speed));
                (_b = this._player) === null || _b === void 0 ? void 0 : _b.pause();
            }
        }
        else {
            console.warn('Android device API is not 23+. Cannot set the playbackRate on lower Android APIs.');
        }
    }
    dispose() {
        return new Promise((resolve, reject) => {
            try {
                if (this._player) {
                    this._player.stop();
                    this._player.reset();
                    this._options = undefined;
                    Application.android.unregisterBroadcastReceiver(android.media.AudioManager.ACTION_AUDIO_BECOMING_NOISY);
                    this._abandonAudioFocus();
                    this.setAudioFocusManager(null);
                }
                resolve(null);
            }
            catch (ex) {
                reject(ex);
            }
        });
    }
    isAudioPlaying() {
        if (this._player) {
            return this._player.isPlaying();
        }
        else {
            return false;
        }
    }
    getAudioTrackDuration() {
        return new Promise((resolve, reject) => {
            try {
                const duration = this._player ? this._player.getDuration() : 0;
                resolve(duration.toString());
            }
            catch (ex) {
                reject(ex);
            }
        });
    }
    _sendEvent(eventName, data) {
        if (this.events) {
            this.events.notify({
                eventName,
                object: this,
                data: data
            });
        }
    }
    _requestAudioFocus() {
        var _a;
        return (_a = this._audioFocusManager) === null || _a === void 0 ? void 0 : _a.requestAudioFocus(this);
    }
    _abandonAudioFocus(preserveMP = false) {
        var _a;
        (_a = this._audioFocusManager) === null || _a === void 0 ? void 0 : _a.abandonAudioFocus(this);
        if (this._mediaPlayer && !preserveMP) {
            this._mediaPlayer.release();
            this._mediaPlayer = undefined;
        }
    }
    _getAndroidContext() {
        let ctx = Application.android.context;
        if (!ctx) {
            ctx = Application.getNativeApplication().getApplicationContext();
        }
        if (ctx === null) {
            setTimeout(() => {
                this._getAndroidContext();
            }, 200);
            return;
        }
        return ctx;
    }
    get _player() {
        if (!this._mediaPlayer && this._options) {
            this._mediaPlayer = new android.media.MediaPlayer();
            this._mediaPlayer.setOnCompletionListener(new android.media.MediaPlayer.OnCompletionListener({
                onCompletion: mp => {
                    if (this._options && this._options.completeCallback) {
                        if (this._options.loop === true) {
                            mp.seekTo(5);
                            mp.start();
                        }
                        this._options.completeCallback({ player: mp });
                    }
                    if (this._options && !this._options.loop) {
                        this._abandonAudioFocus(true);
                    }
                }
            }));
            this._mediaPlayer.setOnErrorListener(new android.media.MediaPlayer.OnErrorListener({
                onError: (player, error, extra) => {
                    if (this._options && this._options.errorCallback) {
                        this._options.errorCallback({ player, error, extra });
                    }
                    this.dispose();
                    return true;
                }
            }));
        }
        return this._mediaPlayer;
    }
    _onAudioFocusChange(data) {
        var _a, _b, _c, _d;
        const focusChange = data.focusChange;
        switch (focusChange) {
            case android.media.AudioManager.AUDIOFOCUS_GAIN:
                if (this._lastPlayerVolume && this._lastPlayerVolume >= 10) {
                    this.volume = 1.0;
                }
                else if (this._lastPlayerVolume) {
                    this.volume = parseFloat('0.' + this._lastPlayerVolume.toString());
                }
                if (this._wasPlaying) {
                    this.resume();
                }
                break;
            case android.media.AudioManager.AUDIOFOCUS_GAIN_TRANSIENT:
                break;
            case android.media.AudioManager.AUDIOFOCUS_LOSS:
                this._wasPlaying = (_b = (_a = this._player) === null || _a === void 0 ? void 0 : _a.isPlaying()) !== null && _b !== void 0 ? _b : false;
                this.pause();
                break;
            case android.media.AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                this._wasPlaying = (_d = (_c = this._player) === null || _c === void 0 ? void 0 : _c.isPlaying()) !== null && _d !== void 0 ? _d : false;
                this.pause();
                break;
            case android.media.AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                this._lastPlayerVolume = this.volume;
                this.volume = 0.2;
                break;
        }
    }
}
